kernel/host/xmicterm/host/xmicterm.cc

```diff
@@ -151,52 +147,54 @@ int main(int argc, char** argv)
     if (info_ptr->info == 0) sleep(1);
   } while (info_ptr->info == 0);
   std::cerr << "host info is at physical address "
-	    << (void*)info_ptr->info << std::endl;
+    << (void*)info_ptr->info << std::endl;
 
   auto info = micmem.access(PhysPtr<HostInfoTable>(info_ptr->info));
   cpu::clflush(info.addr());
   std::cerr << "debug stream is at physical address "
-	    << (void*)info->debugOut << std::endl;
+    << (void*)info->debugOut << std::endl;
 
   auto debugOutChannel = micmem.access(PhysPtr<HostInfoTable::DebugChannel>(info->debugOut));
   PCIeRingConsumer<HostInfoTable::DebugChannel> debugOut(debugOutChannel.addr());
 
-  undo_root(old_uid);
+  // drop root to allow file logger access to user specific directories
+  drop_root();
 
   // collect messages until whole message arrived
-  std::unordered_map<uint16_t, std::string> messages;
+  std::unordered_map<uint16_t, std::stringstream*> messages;
 
-  // log to file
-  file_logger logger;
+  // log to directory debug
+  file_logger logger("debug");
 
   while (true) {
     //typedef HostInfoTable::DebugChannel::handle_t handle_t;
-    /// @todo log messages into a file
     auto handle = debugOut.acquireRecv();
     auto& msg = debugOut.get<DebugMsg>(handle);
 
     auto found = messages.find(msg.vchannel);
-    if (found != messages.end()) { // have to merge messages
-      if (msg.msgbytes < DebugMsg::PAYLOAD) { //output merged messages
-          std::string out(found->second + std::string(msg.data, msg.msgbytes));
-          std::cout << found->first  << "(" << out.length() << "): ";
-          std::cout.write(out.c_str(), out.length());
-          std::cout << std::endl;
-          logger.log(found->first, out);
-          messages.erase(msg.vchannel);
+    if (found != messages.end()) { // a message already there, have to merge them
+      if (msg.msgbytes < DebugMsg::PAYLOAD) { // output merged messages
+        found->second->write(msg.data, msg.msgbytes);
+        auto str = found->second->str();
+        std::cout << found->first  << "(" << str.length() << "): ";
```

Comment:
The length field in the output was a reading help, which was needed before you implemented the message merging. It can be removed now. I think it is also fine to move the std::cout stuff into the logger.